extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/shuffle/DurableStorageOutputChannelFactory.java

```diff
@@ -112,31 +133,28 @@ public OutputChannel openNilChannel(int partitionNumber)
       throw new ISE(
           e,
           "Unable to create empty remote output of workerTask[%s] stage[%d] partition[%d]",
-          workerTaskId,
+          workerNumber,
           stageNumber,
           partitionNumber
       );
     }
   }
 
-  public static String getControllerDirectory(final String controllerTaskId)
-  {
-    return StringUtils.format("controller_%s", IdUtils.validateId("controller task ID", controllerTaskId));
-  }
-
-  public static String getPartitionFileName(
-      final String controllerTaskId,
-      final String workerTaskId,
-      final int stageNumber,
-      final int partitionNumber
-  )
+  /**
+   * Creates a file with name __success and adds the worker's id which has successfully written its outputs. While reading
+   * this file can be used to find out the worker which has written its outputs completely.
+   * Rename operation is not very quick in cloud storage like S3 due to which this alternative
+   * route has been taken
+   */
+  public void createSuccessFile(String taskId) throws IOException
```

Comment:
We should also mention that if the path already exists this operation is a no op.